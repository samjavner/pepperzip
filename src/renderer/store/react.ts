import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
