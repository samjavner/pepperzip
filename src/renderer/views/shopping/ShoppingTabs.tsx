import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import * as actions from "../../app/actions";
import { State } from "../../app/state";
import { Navtab, Navtabs } from "../../components/Navtabs";

function mapStateToProps(state: State) {
    return {
        activePage: state.activeShoppingPage,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {
        onGroceryListsClick: () =>
            dispatch(actions.setActiveShoppingPage("grocery_lists")),
        onInventoryClick: () =>
            dispatch(actions.setActiveShoppingPage("inventory")),
    };
}

type ShoppingTabsProps = ReturnType<typeof mapStateToProps> &
    ReturnType<typeof mapDispatchToProps>;

export const ShoppingTabs: React.SFC<ShoppingTabsProps> = props => (
    <Navtabs>
        <Navtab
            isActive={props.activePage === "grocery_lists"}
            icon="shopping-cart"
            text="Grocery Lists"
            onClick={props.onGroceryListsClick}
        />
        <Navtab
            isActive={props.activePage === "inventory"}
            icon="inbox"
            text="Inventory"
            onClick={props.onInventoryClick}
        />
    </Navtabs>
);

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ShoppingTabs);
