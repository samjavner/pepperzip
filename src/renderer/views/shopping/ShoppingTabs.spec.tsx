import * as React from "react";
import * as renderer from "react-test-renderer";
import { ShoppingTabs } from "./ShoppingTabs";

describe("ShoppingTabs", () => {
    const onGroceryListsClick = () => ({} as any);
    const onInventoryClick = () => ({} as any);

    it("should render with the 'grocery_lists' tab active when the 'grocery_lists' page is active", () => {
        const component = renderer.create(
            <ShoppingTabs
                activePage="grocery_lists"
                onGroceryListsClick={onGroceryListsClick}
                onInventoryClick={onInventoryClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render with the 'inventory' tab active when the 'inventory' page is active", () => {
        const component = renderer.create(
            <ShoppingTabs
                activePage="inventory"
                onGroceryListsClick={onGroceryListsClick}
                onInventoryClick={onInventoryClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should handle click on 'grocery_lists' tab", () => {
        const component = renderer.create(
            <ShoppingTabs
                activePage="grocery_lists"
                onGroceryListsClick={onGroceryListsClick}
                onInventoryClick={onInventoryClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("is-active") !== -1
        ).children[0] as renderer.ReactTestInstance;
        expect(button.props.onClick).toBe(onGroceryListsClick);
    });

    it("should handle click on 'inventory' tab", () => {
        const component = renderer.create(
            <ShoppingTabs
                activePage="inventory"
                onGroceryListsClick={onGroceryListsClick}
                onInventoryClick={onInventoryClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("is-active") !== -1
        ).children[0] as renderer.ReactTestInstance;
        expect(button.props.onClick).toBe(onInventoryClick);
    });
});
