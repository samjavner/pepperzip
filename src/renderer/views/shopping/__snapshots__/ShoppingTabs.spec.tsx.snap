// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingTabs should render with the 'grocery_lists' tab active when the 'grocery_lists' page is active 1`] = `
<div
  className="tabs is-centered"
  style={
    Object {
      "marginBottom": "0px",
    }
  }
>
  <ul>
    <li
      className="is-active"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="icon is-small"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-shopping-cart fa-w-18 "
            data-icon="shopping-cart"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        Grocery Lists
      </a>
    </li>
    <li
      className=""
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="icon is-small"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-box fa-w-16 "
            data-icon="box"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M509.5 184.6L458.9 32.8C452.4 13.2 434.1 0 413.4 0H272v192h238.7c-.4-2.5-.4-5-1.2-7.4zM240 0H98.6c-20.7 0-39 13.2-45.5 32.8L2.5 184.6c-.8 2.4-.8 4.9-1.2 7.4H240V0zM0 224v240c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V224H0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        Inventory
      </a>
    </li>
  </ul>
</div>
`;

exports[`ShoppingTabs should render with the 'inventory' tab active when the 'inventory' page is active 1`] = `
<div
  className="tabs is-centered"
  style={
    Object {
      "marginBottom": "0px",
    }
  }
>
  <ul>
    <li
      className=""
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="icon is-small"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-shopping-cart fa-w-18 "
            data-icon="shopping-cart"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        Grocery Lists
      </a>
    </li>
    <li
      className="is-active"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <span
          className="icon is-small"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-box fa-w-16 "
            data-icon="box"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M509.5 184.6L458.9 32.8C452.4 13.2 434.1 0 413.4 0H272v192h238.7c-.4-2.5-.4-5-1.2-7.4zM240 0H98.6c-20.7 0-39 13.2-45.5 32.8L2.5 184.6c-.8 2.4-.8 4.9-1.2 7.4H240V0zM0 224v240c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V224H0z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
        Inventory
      </a>
    </li>
  </ul>
</div>
`;
